baseline-error-prone/src/main/java/com/palantir/baseline/errorprone/DangerousRecordArrayField.java

```diff
@@ -0,0 +1,99 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.baseline.errorprone;
+
+import com.google.auto.service.AutoService;
+import com.google.errorprone.BugPattern;
+import com.google.errorprone.BugPattern.SeverityLevel;
+import com.google.errorprone.VisitorState;
+import com.google.errorprone.bugpatterns.BugChecker;
+import com.google.errorprone.matchers.Description;
+import com.google.errorprone.matchers.Matcher;
+import com.google.errorprone.matchers.Matchers;
+import com.google.errorprone.util.ASTHelpers;
+import com.sun.source.tree.ClassTree;
+import com.sun.source.tree.MethodTree;
+import com.sun.source.tree.Tree;
+import com.sun.source.tree.VariableTree;
+import com.sun.tools.javac.code.Symbol.ClassSymbol;
+
+/**
+ * Warns that users should not have a {@link java.util.regex.Pattern} as a key to a Set or Map.
+ */
+@AutoService(BugChecker.class)
+@BugPattern(
+        summary = "Record type has an array field and hasn't overridden equals/hashcode. By default array equality"
+                + " performs reference equality only. Consider using an immutable Collection for the field, using"
+                + " Immutables instead of the record, or overriding equals/hashCode in the record.",
+        severity = SeverityLevel.WARNING)
+public final class DangerousRecordArrayField extends BugChecker implements BugChecker.ClassTreeMatcher {
+
+    private static final Matcher<VariableTree> IS_ARRAY_VARIABLE = Matchers.isArrayType();
+    private static final Matcher<MethodTree> NON_TRIVIAL_EQUALS = Matchers.allOf(
+            Matchers.equalsMethodDeclaration(),
+            Matchers.not(Matchers.singleStatementReturnMatcher(Matchers.instanceEqualsInvocation())));
+    private static final Matcher<MethodTree> NON_TRIVIAL_HASHCODE = Matchers.allOf(
+            Matchers.hashCodeMethodDeclaration(),
+            Matchers.not(Matchers.singleStatementReturnMatcher(Matchers.instanceHashCodeInvocation())));
+
+    @Override
+    public Description matchClass(ClassTree classTree, VisitorState state) {
+        ClassSymbol classSymbol = ASTHelpers.getSymbol(classTree);
+        if (!ASTHelpers.isRecord(classSymbol)) {
+            return Description.NO_MATCH;
+        }
+        if (!hasArrayField(classTree, state)) {
+            return Description.NO_MATCH;
+        }
+        if (hasNonTrivialEqualsAndHashCode(classTree, state)) {
+            return Description.NO_MATCH;
+        }
+
+        return buildDescription(classTree).build();
```

Comment:
Perhaps we should suggest replacing `T[]` with `ImmutableList<T>` (with an automated refactor?)

Reply:
I think I disagree for two reasons:
 - Going from e.g. `byte[]` to `ImmutableList<Byte>` has _a lot_ of overheads (both memory & conversion). So I'm not a fan of automatically refactoring. These can end up being quite big (e.g. imagine if an file-serving service ever gets this as an automatic refactor).
 - Restricting to `ImmutableList` (as opposed to `List`), while ensuring the typesystem helps you, also creates a bunch of code noise & extra allocations for very little gain given our programming culture. I don't really ever see people mutating the type of value that get inserted into records.

Reply:
Note that `Arrays.asList(Byte[])` ~> `List<Byte>` & `Arrays.asList(byte[])` ~> `List<byte[]>`; to my knowledge there's no jdk-included zero-copy way to transform a byte[] to a List<Byte>.

If we assume guava we could recommend `Bytes.asList` (but that only solves the problem for byte[], not for other types).

Reply:
Idea isn't to automatically do the refactor, but to add a suggestion inline to nudge folks toward best practices. I've found it helpful when error-prone checks fail with both a human readable description as well as the text "did you mean `foo.bar(baz)`?"

> Going from e.g. `byte[]` to `ImmutableList<Byte>`

We can provide other specific recommendations for binary data like the conjure `Bytes` type.

> Restricting to ImmutableList (as opposed to List)

I'm not suggesting that we restrict to ImmutableList, rather that it makes sense as a default recommendation.

The idea here is to use the check failure opportunity to share as much context as we can to help folks make the best, most informed decisions.

> I don't really ever see people mutating the type of value that get inserted into records.

It's very hard to see, this sort of thing often becomes a bug when it occurs on the original list after the record has been created and passed elsewhere. I don't think our default stance should be to allow subtle mutability. I agree wholeheartedly that it's often necessary for performance, but I think that should be the exception rather than the rule because debugging issues resulting from unexpected mutations is a nightmare.

Reply:
Note that conjure's `Bytes` type still does a full copy, which is not ideal when the goal was to return a tuple (a classic use-case of `record`).

Reply:
I'm happy to see if I can auto-generate an equality/hashCode implementation.

Reply:
> Note that conjure's Bytes type still does a full copy, which is not ideal when the goal was to return a tuple (a classic use-case of record).

In such cases the equals and hashCode methods are seldom used either, no?

Reply:
Correct; we could also try to detect the usage of equals/hashCode on such records & complain there, but I figured we'd start somewhere.

Separately, in the project that made me realize the issue, the only place where equality was checked was in the unit tests.

Reply:
Agree it's best to flag the definition rather than call sites, detecting collections which eventually rely on hashCode/equals is tricky business. Happy to merge this as is, I think automatically adding equals/hashCode to records would be great, especially any time we migrate existing code into one of our codebases, but I don't want to block on that. Opens the door to other potential checks to prevent bit-rot, for example ensuring all fields are taken into account in overridden equality and hashCode methods, so that new element additions don't violate the contract.

Reply:
Yeah, 100% agree on the equality/hashCode validity question. Will try to think of ways of doing this, but this is a little far out of my comfort zone w/ error-prone.